\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={MSMB-Chapter2-Statistical Modelling},
            pdfauthor={Aleeza Gerstein},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\providecommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{MSMB-Chapter2-Statistical Modelling}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
    \author{Aleeza Gerstein}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
      \predate{\centering\large\emph}
  \postdate{\par}
    \date{2019-09-16}

\usepackage{color}
\usepackage{mathclap}

\begin{document}
\maketitle

\section{Chapter 2: Statistical
Modeling}\label{chapter-2-statistical-modeling}

\subsection{A simple example of statistical
modelling}\label{a-simple-example-of-statistical-modelling}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{e99 =}\StringTok{ }\KeywordTok{rpois}\NormalTok{(}\DecValTok{99}\NormalTok{,}\DataTypeTok{lambda=}\FloatTok{0.5}\NormalTok{)}
\KeywordTok{barplot}\NormalTok{(}\KeywordTok{table}\NormalTok{(e99), }\DataTypeTok{space =} \FloatTok{0.8}\NormalTok{, }\DataTypeTok{col =} \StringTok{"chartreuse4"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{MSMB-Chapter2_files/figure-latex/barplot_pois-1.pdf}
\caption{The observed distribution of the epitope data without the
outlier.}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(}\StringTok{"vcd"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Loading required package: grid
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gf1 =}\StringTok{ }\KeywordTok{goodfit}\NormalTok{(e99, }\StringTok{"poisson"}\NormalTok{)}
\KeywordTok{rootogram}\NormalTok{(gf1, }\DataTypeTok{xlab =} \StringTok{""}\NormalTok{, }\DataTypeTok{rect_gp =} \KeywordTok{gpar}\NormalTok{(}\DataTypeTok{fill =} \StringTok{"chartreuse4"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{MSMB-Chapter2_files/figure-latex/stat-rooto-1.pdf}
\caption{Rootogram showing the square root of the theoretical values as
red dots and the square root of the observed frequencies as drop down
rectangles. (We'll see a bit below how the \texttt{goodfit} function
decided which \(\lambda\) to use.)}
\end{figure}

\textcolor{blue}{We will learn later what this is?}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textcolor{red}{Question 1: To calibrate what such a plot looks liek a known poisson variable, use `rpois` and $\gamma$}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\subsection{----RootogramPoisson, fig.width= 3, fig.height =
3.5--------------------}\label{rootogrampoisson-fig.width-3-fig.height-3.5}

simp = rpois(100, lambda = 0.05) gf2 = goodfit(simp, ``poisson'')
rootogram(gf2, xlab = ``'')

\subsection{----table100------------------------------------------------------------}\label{table100}

table(e100)

\subsection{----table3--------------------------------------------------------------}\label{table3}

table(rpois(100, 3))

\subsection{----table101, echo =
FALSE----------------------------------------------}\label{table101-echo-false-}

counts = table(e100) stopifnot(identical(names(counts), c(``0'', ``1'',
``2'', ``7'')), all(counts==c(58, 34, 7, 1)))

\subsection{----poism3--------------------------------------------------------------}\label{poism3}

prod(dpois(c(0, 1, 2, 7), lambda = 3) \^{} (c(58, 34, 7, 1)))

\subsection{----anspois-------------------------------------------------------------}\label{anspois-}

prod(dpois(c(0, 1, 2, 7), lambda = 0.4) \^{} (c(58, 34, 7, 1)))

\subsection{----functionll----------------------------------------------------------}\label{functionll-}

loglikelihood = function(lambda, data = e100) \{ sum(log(dpois(data,
lambda))) \}

\subsection{\texorpdfstring{----chap2-r-poislikel-1, fig.keep = `high',
fig.cap = ``The red curve is the log-likelihood function. The vertical
line shows the value of \texttt{m} (the mean) and the horizontal line
the log-likelihood of \texttt{m}. It looks like \texttt{m} maximizes the
likelihood.'',
fig.width=3.5----}{----chap2-r-poislikel-1, fig.keep = high, fig.cap = The red curve is the log-likelihood function. The vertical line shows the value of m (the mean) and the horizontal line the log-likelihood of m. It looks like m maximizes the likelihood., fig.width=3.5----}}\label{chap2-r-poislikel-1-fig.keep-high-fig.cap-the-red-curve-is-the-log-likelihood-function.-the-vertical-line-shows-the-value-of-m-the-mean-and-the-horizontal-line-the-log-likelihood-of-m.-it-looks-like-m-maximizes-the-likelihood.-fig.width3.5-}

lambdas = seq(0.05, 0.95, length = 100) loglik = vapply(lambdas,
loglikelihood, numeric(1)) plot(lambdas, loglik, type = ``l'', col =
``red'', ylab = ``'', lwd = 2, xlab = expression(lambda)) m0 =
mean(e100) abline(v = m0, col = ``blue'', lwd = 2) abline(h =
loglikelihood(m0), col = ``purple'', lwd = 2) m0

\subsection{----gfpoisson-----------------------------------------------------------}\label{gfpoisson}

gf = goodfit(e100, ``poisson'') names(gf) gf\$par

\subsection{----colorblind, echo =
FALSE--------------------------------------------}\label{colorblind-echo-false}

cb = c(rep(0, 110), rep(1, 10))

\subsection{----cb------------------------------------------------------------------}\label{cb}

table(cb)

\subsection{\texorpdfstring{----likely1, fig.keep = `high', fig.cap =
``Plot of the likelihood as a function of the probabilities. The
likelihood is a function on \([0, 1]\); here we have zoomed into the
range of \([(ref:likely1-1), (ref:likely1-2)]\), as the likelihood is
practically zero for larger values of \(p\).'', fig.width =
4----}{----likely1, fig.keep = high, fig.cap = Plot of the likelihood as a function of the probabilities. The likelihood is a function on {[}0, 1{]}; here we have zoomed into the range of {[}(ref:likely1-1), (ref:likely1-2){]}, as the likelihood is practically zero for larger values of p., fig.width = 4----}}\label{likely1-fig.keep-high-fig.cap-plot-of-the-likelihood-as-a-function-of-the-probabilities.-the-likelihood-is-a-function-on-0-1-here-we-have-zoomed-into-the-range-of-reflikely1-1-reflikely1-2-as-the-likelihood-is-practically-zero-for-larger-values-of-p.-fig.width-4-}

probs = seq(0, 0.3, by = 0.005) likelihood = dbinom(sum(cb), prob =
probs, size = length(cb)) plot(probs, likelihood, pch = 16, xlab =
``probability of success'', ylab = ``likelihood'', cex=0.6)
probs{[}which.max(likelihood){]}

\subsection{----check, echo =
FALSE-------------------------------------------------}\label{check-echo-false-}

stopifnot(abs(probs{[}which.max(likelihood){]}-1/12) \textless{}
diff(probs{[}1:2{]}))

\subsection{----loglike1------------------------------------------------------------}\label{loglike1}

loglikelihood = function(theta, n = 300, k = 40) \{ 115 + k * log(theta)
+ (n - k) * log(1 - theta) \}

\subsection{\texorpdfstring{----chap2-r-loglikelihood-1, fig.keep =
`high', fig.cap = ``Plot of the log likelihood function for \(n=300\)
and \(y=40\).'', fig.width = 3, fig.height =
3.2----}{----chap2-r-loglikelihood-1, fig.keep = high, fig.cap = Plot of the log likelihood function for n=300 and y=40., fig.width = 3, fig.height = 3.2----}}\label{chap2-r-loglikelihood-1-fig.keep-high-fig.cap-plot-of-the-log-likelihood-function-for-n300-and-y40.-fig.width-3-fig.height-3.2-}

thetas = seq(0, 1, by = 0.001) plot(thetas, loglikelihood(thetas), xlab
= expression(theta), ylab = expression(paste(``log f('', theta, "
\textbar{} y)``)),type =''l``)

\subsection{----staph---------------------------------------------------------------}\label{staph}

library(``Biostrings'') staph =
readDNAStringSet(``../data/staphsequence.ffn.txt'', ``fasta'')

\subsection{----firstgenestaph------------------------------------------------------}\label{firstgenestaph}

staph{[}1{]} letterFrequency(staph{[}{[}1{]}{]}, letters = ``ACGT'', OR
= 0)

\subsection{----compareprop---------------------------------------------------------}\label{compareprop}

letterFrq = vapply(staph, letterFrequency, FUN.VALUE = numeric(4),
letters = ``ACGT'', OR = 0) colnames(letterFrq) = paste0(``gene'',
seq(along = staph)) tab10 = letterFrq{[}, 1:10{]} computeProportions =
function(x) \{ x/sum(x) \} prop10 = apply(tab10, 2, computeProportions)
round(prop10, digits = 2) p0 = rowMeans(prop10) p0

\subsection{----outerex-------------------------------------------------------------}\label{outerex-}

cs = colSums(tab10) cs expectedtab10 = outer(p0, cs, FUN = ``*``)
round(expectedtab10)

\subsection{----genrandomtabs-------------------------------------------------------}\label{genrandomtabs-}

randomtab10 = sapply(cs, function(s) \{ rmultinom(1, s, p0) \} )
all(colSums(randomtab10) == cs)

\subsection{----assertgenrandomtabs, echo =
FALSE-----------------------------------}\label{assertgenrandomtabs-echo-false}

stopifnot(all(colSums(randomtab10) == cs))

\subsection{\texorpdfstring{----chap2-r-quant12-1, fig.keep = `high',
fig.cap = ``Histogram of \texttt{simulstat}. The value of \texttt{S1} is
marked by the vertical red line, those of the 0.95 and 0.99 quantiles
(see next section) by the dotted lines.'', fig.width = 4, fig.height =
3.5----}{----chap2-r-quant12-1, fig.keep = high, fig.cap = Histogram of simulstat. The value of S1 is marked by the vertical red line, those of the 0.95 and 0.99 quantiles (see next section) by the dotted lines., fig.width = 4, fig.height = 3.5----}}\label{chap2-r-quant12-1-fig.keep-high-fig.cap-histogram-of-simulstat.-the-value-of-s1-is-marked-by-the-vertical-red-line-those-of-the-0.95-and-0.99-quantiles-see-next-section-by-the-dotted-lines.-fig.width-4-fig.height-3.5-}

stat = function(obsvd, exptd = 20 * pvec) \{ sum((obsvd - exptd)\^{}2 /
exptd) \} B = 1000 simulstat = replicate(B, \{ randomtab10 = sapply(cs,
function(s) \{ rmultinom(1, s, p0) \}) stat(randomtab10, expectedtab10)
\}) S1 = stat(tab10, expectedtab10) sum(simulstat \textgreater{}= S1)

hist(simulstat, col = ``lavender'', breaks = seq(0, 75, length.out=50))
abline(v = S1, col = ``red'') abline(v = quantile(simulstat, probs =
c(0.95, 0.99)), col = c(``darkgreen'', ``blue''), lty = 2)

\subsection{----checksimulstat, echo =
FALSE----------------------------------------}\label{checksimulstat-echo-false-}

stopifnot(max(simulstat)\textless{}75, S1\textless{}75)

\subsection{\texorpdfstring{----quantiles3, results =
``hide''----------------------------------------}{----quantiles3, results = hide----------------------------------------}}\label{quantiles3-results-hide-}

qs = ppoints(100) quantile(simulstat, qs) quantile(qchisq(qs, df = 30),
qs)

\subsection{\texorpdfstring{----chap2-r-qqplot3-1, fig.keep = `high',
fig.cap = ``Our simulated statistic's distribution compared to
\(\\chi_{30}^2\) using a QQ-plot, which shows the theoretical
\textbf{quantiles} for the \(\\chi^2_{30}\) distribution on the
horizontal axis and the sampled ones on the vertical axis.'', fig.width
= 3.4, fig.height =
4----}{----chap2-r-qqplot3-1, fig.keep = high, fig.cap = Our simulated statistic's distribution compared to \textbackslash{}\textbackslash{}chi\_\{30\}\^{}2 using a QQ-plot, which shows the theoretical quantiles for the \textbackslash{}\textbackslash{}chi\^{}2\_\{30\} distribution on the horizontal axis and the sampled ones on the vertical axis., fig.width = 3.4, fig.height = 4----}}\label{chap2-r-qqplot3-1-fig.keep-high-fig.cap-our-simulated-statistics-distribution-compared-to-chi_302-using-a-qq-plot-which-shows-the-theoretical-quantiles-for-the-chi2_30-distribution-on-the-horizontal-axis-and-the-sampled-ones-on-the-vertical-axis.-fig.width-3.4-fig.height-4-}

qqplot(qchisq(ppoints(B), df = 30), simulstat, main = ``'', xlab =
expression(chi{[}nu==30{]}\^{}2), asp = 1, cex = 0.5, pch = 16) abline(a
= 0, b = 1, col = ``red'')

\subsection{----pvalueBias----------------------------------------------------------}\label{pvaluebias-}

1 - pchisq(S1, df = 30)

\subsection{\texorpdfstring{---- ChargaffColdSpring, eval = TRUE, echo =
FALSE, fig.show = `hold', fig.keep =
`high'----}{---- ChargaffColdSpring, eval = TRUE, echo = FALSE, fig.show = hold, fig.keep = high----}}\label{chargaffcoldspring-eval-true-echo-false-fig.show-hold-fig.keep-high-}

knitr::include\_graphics(c(`images/ChargaffColdSpring.png'))

\subsection{----Chargaff------------------------------------------------------------}\label{chargaff}

load(``../data/ChargaffTable.RData'') ChargaffTable

\subsection{\texorpdfstring{----ChargaffBars, fig.keep = `high', fig.cap
= ``Barplots for the different rows in \texttt{ChargaffTable}. Can you
spot the pattern?'', fig.margin = FALSE, echo = FALSE, fig.width = 7,
fig.height =
3.4----}{----ChargaffBars, fig.keep = high, fig.cap = Barplots for the different rows in ChargaffTable. Can you spot the pattern?, fig.margin = FALSE, echo = FALSE, fig.width = 7, fig.height = 3.4----}}\label{chargaffbars-fig.keep-high-fig.cap-barplots-for-the-different-rows-in-chargafftable.-can-you-spot-the-pattern-fig.margin-false-echo-false-fig.width-7-fig.height-3.4-}

stopifnot(nrow(ChargaffTable) == 8) mycolors = c(``chocolate'',
``aquamarine4'', ``cadetblue4'', ``coral3'',
``chartreuse4'',``darkgoldenrod4'',``darkcyan'',``brown4'')
par(mfrow=c(2, 4), mai = c(0, 0.7, 0.7, 0)) for (i in 1:8) \{ cbp =
barplot(ChargaffTable{[}i, {]}, horiz = TRUE, axes = FALSE, axisnames =
FALSE, col = mycolors{[}i{]}) ax = axis(3, las = 2, labels = FALSE, col
= mycolors{[}i{]}, cex = 0.5, at = c(0, 10, 20)) mtext(side = 3, at =
ax, text = paste(ax), col = mycolors{[}i{]}, line = 0, las = 1, cex =
0.9) mtext(side = 2, at = cbp, text = colnames(ChargaffTable), col =
mycolors{[}i{]}, line = 0, las = 2, cex = 1)
title(paste(rownames(ChargaffTable){[}i{]}), col = mycolors{[}i{]}, cex
= 1.1) \}

\subsection{\texorpdfstring{----chap2-r-permstatChf-1, fig.keep =
`high', fig.cap = ``Histogram of our statistic \texttt{statChf} computed
from simulations using per-row permutations of the columns. The value it
yields for the observed data is shown by the red line.'', fig.width =
3.2, fig.height =
3----}{----chap2-r-permstatChf-1, fig.keep = high, fig.cap = Histogram of our statistic statChf computed from simulations using per-row permutations of the columns. The value it yields for the observed data is shown by the red line., fig.width = 3.2, fig.height = 3----}}\label{chap2-r-permstatchf-1-fig.keep-high-fig.cap-histogram-of-our-statistic-statchf-computed-from-simulations-using-per-row-permutations-of-the-columns.-the-value-it-yields-for-the-observed-data-is-shown-by-the-red-line.-fig.width-3.2-fig.height-3-}

statChf = function(x)\{ sum((x{[}, ``C''{]} - x{[}, ``G''{]})\^{}2 +
(x{[}, ``A''{]} - x{[}, ``T''{]})\^{}2) \} chfstat =
statChf(ChargaffTable) permstat = replicate(100000, \{ permuted =
t(apply(ChargaffTable, 1, sample)) colnames(permuted) =
colnames(ChargaffTable) statChf(permuted) \}) pChf = mean(permstat
\textless{}= chfstat) pChf hist(permstat, breaks = 100, main = ``'', col
= ``lavender'') abline(v = chfstat, lwd = 2, col = ``red'')

\subsection{----vcdHC---------------------------------------------------------------}\label{vcdhc}

HairEyeColor{[},, ``Female''{]}

\subsection{----answerHC------------------------------------------------------------}\label{answerhc}

str(HairEyeColor) ? HairEyeColor

\subsection{----Deuto---------------------------------------------------------------}\label{deuto}

load(``../data/Deuteranopia.RData'') Deuteranopia

\subsection{----chisq.test.Deuteranopia---------------------------------------------}\label{chisq.test.deuteranopia}

chisq.test(Deuteranopia)

\subsection{\texorpdfstring{----chap2-r-HardyWeinberg-1, fig.keep =
`high', fig.cap = ``Plot of the log-likelihood for the
(ref:chap2-r-HardyWeinberg-1-1)
data.''----}{----chap2-r-HardyWeinberg-1, fig.keep = high, fig.cap = Plot of the log-likelihood for the (ref:chap2-r-HardyWeinberg-1-1) data.----}}\label{chap2-r-hardyweinberg-1-fig.keep-high-fig.cap-plot-of-the-log-likelihood-for-the-refchap2-r-hardyweinberg-1-1-data.-}

library(``HardyWeinberg'') data(``Mourant'') Mourant{[}214:216,{]} nMM =
Mourant\(MM[216] nMN = Mourant\)MN{[}216{]} nNN = Mourant\$NN{[}216{]}
loglik = function(p, q = 1 - p) \{ 2 * nMM * log(p) + nMN *
log(2\emph{p}q) + 2 * nNN * log(q) \} xv = seq(0.01, 0.99, by = 0.01) yv
= loglik(xv) plot(x = xv, y = yv, type = ``l'', lwd = 2, xlab = ``p'',
ylab = ``log-likelihood'') imax = which.max(yv) abline(v = xv{[}imax{]},
h = yv{[}imax{]}, lwd = 1.5, col = ``blue'') abline(h = yv{[}imax{]},
lwd = 1.5, col = ``purple'')

\subsection{----phat----------------------------------------------------------------}\label{phat-}

phat = af(c(nMM, nMN, nNN)) phat pMM = phat\^{}2 qhat = 1 - phat

\subsection{----hweq----------------------------------------------------------------}\label{hweq-}

pHW = c(MM = phat\^{}2, MN = 2\emph{phat}qhat, NN = qhat\^{}2)
sum(c(nMM, nMN, nNN)) * pHW

\subsection{\texorpdfstring{----HWtern, fig.keep = `high', fig.cap =
``This \textbf{de Finetti plot} shows the points as barycenters of the
three genotypes using the frequencies as weights on each of the corners
of the triangle. The Hardy-Weinberg model is the red curve, the
acceptance region is between the two purple lines. We see that the US is
the furthest from being in HW equilibrium.'', fig.margin = FALSE,
message = FALSE, warning = FALSE, fig.width = 3.4, fig.height = 3.4,
results = FALSE, echo =
-1----}{----HWtern, fig.keep = high, fig.cap = This de Finetti plot shows the points as barycenters of the three genotypes using the frequencies as weights on each of the corners of the triangle. The Hardy-Weinberg model is the red curve, the acceptance region is between the two purple lines. We see that the US is the furthest from being in HW equilibrium., fig.margin = FALSE, message = FALSE, warning = FALSE, fig.width = 3.4, fig.height = 3.4, results = FALSE, echo = -1----}}\label{hwtern-fig.keep-high-fig.cap-this-de-finetti-plot-shows-the-points-as-barycenters-of-the-three-genotypes-using-the-frequencies-as-weights-on-each-of-the-corners-of-the-triangle.-the-hardy-weinberg-model-is-the-red-curve-the-acceptance-region-is-between-the-two-purple-lines.-we-see-that-the-us-is-the-furthest-from-being-in-hw-equilibrium.-fig.margin-false-message-false-warning-false-fig.width-3.4-fig.height-3.4-results-false-echo--1-}

par(mai = rep(0.1, 4)) pops = c(1, 69, 128, 148, 192)
genotypeFrequencies = as.matrix(Mourant{[}, c(``MM'', ``MN'',
``NN''){]}) HWTernaryPlot(genotypeFrequencies{[}pops, {]}, markerlab =
Mourant\$Country{[}pops{]}, alpha = 0.0001, curvecols = c(``red'',
rep(``purple'', 4)), mcex = 0.75, vertex.cex = 1)

\subsection{----quesTern, echo =
-1-------------------------------------------------}\label{questern-echo--1-}

HWTernaryPlot(genotypeFrequencies{[}pops, {]}, markerlab =
Mourant\$Country{[}pops{]}, alpha = 0.0001, curvecols = c(``red'',
rep(``purple'', 4)), mcex = 0.75, vertex.cex = 1)
HWTernaryPlot(genotypeFrequencies{[}-pops, {]}, alpha = 0.0001, newframe
= FALSE, cex = 0.5)

\subsection{----newMNdata-----------------------------------------------------------}\label{newmndata}

newgf = round(genotypeFrequencies / 50) HWTernaryPlot(newgf{[}pops, {]},
markerlab = Mourant\$Country{[}pops{]}, alpha = 0.0001, curvecols =
c(``red'', rep(``purple'', 4)), mcex = 0.75, vertex.cex = 1)

\subsection{\texorpdfstring{----chap2-r-seqlogo-1, fig.keep = `high',
fig.cap = ``Here is a diagram called a sequence logo for the position
dependent multinomial used to model the Kozak motif. It codifies the
amount of variation in each of the positions on a log scale. The large
letters represent positions where there is no uncertainty about which
nucleotide occurs.'', fig.margin = FALSE, fig.height=5,
fig.width=5----}{----chap2-r-seqlogo-1, fig.keep = high, fig.cap = Here is a diagram called a sequence logo for the position dependent multinomial used to model the Kozak motif. It codifies the amount of variation in each of the positions on a log scale. The large letters represent positions where there is no uncertainty about which nucleotide occurs., fig.margin = FALSE, fig.height=5, fig.width=5----}}\label{chap2-r-seqlogo-1-fig.keep-high-fig.cap-here-is-a-diagram-called-a-sequence-logo-for-the-position-dependent-multinomial-used-to-model-the-kozak-motif.-it-codifies-the-amount-of-variation-in-each-of-the-positions-on-a-log-scale.-the-large-letters-represent-positions-where-there-is-no-uncertainty-about-which-nucleotide-occurs.-fig.margin-false-fig.height5-fig.width5-}

library(``seqLogo'') load(``../data/kozak.RData'') kozak pwm =
makePWM(kozak) seqLogo(pwm, ic.scale = FALSE)

\subsection{----4stateMC, echo =
FALSE----------------------------------------------}\label{statemc-echo-false-}

library(``markovchain'') library(``igraph'') sequence = toupper(c(``a'',
``c'', ``a'', ``c'', ``g'', ``t'', ``t'', ``t'', ``t'', ``c'', ``c'',
``a'', ``c'', ``g'', ``t'', ``a'',
``c'',``c'',``c'',``a'',``a'',``a'',``t'',``a'',
``c'',``g'',``g'',``c'',``a'',``t'',``g'',``t'',``g'',``t'',``g'',``a'',``g'',``c'',``t'',``g''))
mcFit = markovchainFit(data = sequence) MCgraph =
markovchain:::.getNet(mcFit\(estimate, round = TRUE) edgelab = round(E(MCgraph)\)weight
/ 100, 1)

\subsection{\texorpdfstring{----statsfourstateMC, fig.keep = `high',
fig.cap = ``Visualisation of a 4-state Markov chain. The probability of
each possible digram (e.\textbackslash{},g., CA) is given by the weight
of the edge between the corresponding nodes. So for instance, the
probability of CA is given by the edge C\(\\to\) A. We'll see in Chapter
\textbackslash{}@ref(Chap:Images) how to use \textbf{R} packages to draw
these type of network graphs.'', echo = FALSE, fig.width = 4, fig.height
=
3.5----}{----statsfourstateMC, fig.keep = high, fig.cap = Visualisation of a 4-state Markov chain. The probability of each possible digram (e.\textbackslash{},g., CA) is given by the weight of the edge between the corresponding nodes. So for instance, the probability of CA is given by the edge C\textbackslash{}\textbackslash{}to A. We'll see in Chapter \textbackslash{}@ref(Chap:Images) how to use R packages to draw these type of network graphs., echo = FALSE, fig.width = 4, fig.height = 3.5----}}\label{statsfourstatemc-fig.keep-high-fig.cap-visualisation-of-a-4-state-markov-chain.-the-probability-of-each-possible-digram-e.g.-ca-is-given-by-the-weight-of-the-edge-between-the-corresponding-nodes.-so-for-instance-the-probability-of-ca-is-given-by-the-edge-cto-a.-well-see-in-chapter-refchapimages-how-to-use-r-packages-to-draw-these-type-of-network-graphs.-echo-false-fig.width-4-fig.height-3.5-}

par(mai=c(0,0,0,0)) \#plot.igraph(MCgraph, edge.label = edgelab, width =
2, edge.arrow.width = 1.5, \# vertex.size = 40, xlim = c(-1, 1.25))
plot.igraph(MCgraph, edge.label = edgelab, \#edge.arrow.width = 1.5,
vertex.size = 40, xlim = c(-1, 1.25))

\subsection{\texorpdfstring{---- FreqBayes-turtles, eval = TRUE, echo =
FALSE, fig.show = `hold', fig.keep = `high', fig.cap = ``Turtles all the
way down. Bayesian modeling of the uncertainty of the parameter of a
distribution is done by using a random variable whose distribution may
depend on parameters whose uncertainty can be modeled as a random
variable; these are called hierarchical
models.''----}{---- FreqBayes-turtles, eval = TRUE, echo = FALSE, fig.show = hold, fig.keep = high, fig.cap = Turtles all the way down. Bayesian modeling of the uncertainty of the parameter of a distribution is done by using a random variable whose distribution may depend on parameters whose uncertainty can be modeled as a random variable; these are called hierarchical models.----}}\label{freqbayes-turtles-eval-true-echo-false-fig.show-hold-fig.keep-high-fig.cap-turtles-all-the-way-down.-bayesian-modeling-of-the-uncertainty-of-the-parameter-of-a-distribution-is-done-by-using-a-random-variable-whose-distribution-may-depend-on-parameters-whose-uncertainty-can-be-modeled-as-a-random-variable-these-are-called-hierarchical-models.-}

knitr::include\_graphics(c(`images/turtlesalltheway.png'))

\subsection{\texorpdfstring{---- STRDefinition, fig.margin = FALSE, eval
= TRUE, echo = FALSE, fig.show = `hold', fig.keep = `high', fig.cap =
``A short tandem repeat (STR) in DNA occurs when a pattern of two or
more nucleotides is repeated and the repeated sequences are directly
adjacent to each other. An STR is also known as a microsatellite. The
pattern can range in length from 2 to 13 nucleotides, and the number of
repeats is highly variable across individuals. STR numbers can be used
as genetic signatures, called
haplotypes.''----}{---- STRDefinition, fig.margin = FALSE, eval = TRUE, echo = FALSE, fig.show = hold, fig.keep = high, fig.cap = A short tandem repeat (STR) in DNA occurs when a pattern of two or more nucleotides is repeated and the repeated sequences are directly adjacent to each other. An STR is also known as a microsatellite. The pattern can range in length from 2 to 13 nucleotides, and the number of repeats is highly variable across individuals. STR numbers can be used as genetic signatures, called haplotypes.----}}\label{strdefinition-fig.margin-false-eval-true-echo-false-fig.show-hold-fig.keep-high-fig.cap-a-short-tandem-repeat-str-in-dna-occurs-when-a-pattern-of-two-or-more-nucleotides-is-repeated-and-the-repeated-sequences-are-directly-adjacent-to-each-other.-an-str-is-also-known-as-a-microsatellite.-the-pattern-can-range-in-length-from-2-to-13-nucleotides-and-the-number-of-repeats-is-highly-variable-across-individuals.-str-numbers-can-be-used-as-genetic-signatures-called-haplotypes.-}

knitr::include\_graphics(c(`images/STRDefinition.png'))

\subsection{----haplo6--------------------------------------------------------------}\label{haplo6}

haplo6=read.table(``../data/haplotype6.txt'',header = TRUE) haplo6

\subsection{\texorpdfstring{----histobeta2, fig.keep = `high', fig.cap =
``Beta distributions with \(\\alpha=10,20,50\) and \(\\beta=30,60,150\)
used as a \{prior\} for a probability of success. These three
distributions have the same mean (\(\\frac{\\alpha}{\\alpha +\\beta}\)),
but different concentrations around the mean.'', echo = FALSE, fig.width
= 3.5, fig.height =
3.5----}{----histobeta2, fig.keep = high, fig.cap = Beta distributions with \textbackslash{}\textbackslash{}alpha=10,20,50 and \textbackslash{}\textbackslash{}beta=30,60,150 used as a \{prior\} for a probability of success. These three distributions have the same mean (\textbackslash{}\textbackslash{}frac\{\textbackslash{}\textbackslash{}alpha\}\{\textbackslash{}\textbackslash{}alpha +\textbackslash{}\textbackslash{}beta\}), but different concentrations around the mean., echo = FALSE, fig.width = 3.5, fig.height = 3.5----}}\label{histobeta2-fig.keep-high-fig.cap-beta-distributions-with-alpha102050-and-beta3060150-used-as-a-prior-for-a-probability-of-success.-these-three-distributions-have-the-same-mean-fracalphaalpha-beta-but-different-concentrations-around-the-mean.-echo-false-fig.width-3.5-fig.height-3.5-}

theta = thetas{[}1:500{]} dfbetas = data.frame(theta, db1=
dbeta(theta,10,30), db2 = dbeta(theta, 20, 60), db3 = dbeta(theta, 50,
150)) require(reshape2) datalong = melt(dfbetas, id=``theta'')
ggplot(datalong) + geom\_line(aes(x = theta,y=value,colour=variable)) +
theme(legend.title=element\_blank()) + geom\_vline(aes(xintercept=0.25),
colour=``\#990000'', linetype=``dashed'')+ scale\_colour\_discrete(name
=``Prior'', labels=c(``B(10,30)'', ``B(20,60)'',``B(50,150)''))

\subsection{\texorpdfstring{----chap2-r-histmarginal-1, fig.keep =
`high', fig.cap = ``Marginal Distribution of \(Y\).'', fig.width = 3.5,
fig.height =
3.5----}{----chap2-r-histmarginal-1, fig.keep = high, fig.cap = Marginal Distribution of Y., fig.width = 3.5, fig.height = 3.5----}}\label{chap2-r-histmarginal-1-fig.keep-high-fig.cap-marginal-distribution-of-y.-fig.width-3.5-fig.height-3.5-}

rtheta = rbeta(100000, 50, 350) y = vapply(rtheta, function(th) \{
rbinom(1, prob = th, size = 300) \}, numeric(1)) hist(y, breaks = 50,
col = ``orange'', main = ``'', xlab = ``'')

\subsection{----freqquesvectorize, echo =
FALSE-------------------------------------}\label{freqquesvectorize-echo-false-}

set.seed(0xbebe) .w1 = vapply(rtheta, function(th) rbinom(1, prob = th,
size = 300), integer(1)) set.seed(0xbebe) .w2 = rbinom(length(rtheta),
rtheta, size = 300) stopifnot(identical(.w1, .w2))

\subsection{\texorpdfstring{----chap2-r-densityposterior-1, fig.keep =
`high', fig.cap = ``Only choosing the values of the distribution with
\(Y=40\) gives the posterior distribution of \(\\theta\). The histogram
(green) shows the simulated values for the posteriror distribution, the
line the theoretical density of a beta distribution with the theoretical
posterior parameters.'', fig.width = 3.5, fig.height =
3.5----}{----chap2-r-densityposterior-1, fig.keep = high, fig.cap = Only choosing the values of the distribution with Y=40 gives the posterior distribution of \textbackslash{}\textbackslash{}theta. The histogram (green) shows the simulated values for the posteriror distribution, the line the theoretical density of a beta distribution with the theoretical posterior parameters., fig.width = 3.5, fig.height = 3.5----}}\label{chap2-r-densityposterior-1-fig.keep-high-fig.cap-only-choosing-the-values-of-the-distribution-with-y40-gives-the-posterior-distribution-of-theta.-the-histogram-green-shows-the-simulated-values-for-the-posteriror-distribution-the-line-the-theoretical-density-of-a-beta-distribution-with-the-theoretical-posterior-parameters.-fig.width-3.5-fig.height-3.5-}

thetaPostEmp = rtheta{[} y == 40 {]} hist(thetaPostEmp, breaks = 40, col
= ``chartreuse4'', main = ``'', probability = TRUE, xlab =
expression(``posterior''\textasciitilde{}theta)) densPostTheory =
dbeta(thetas, 90, 610) lines(thetas, densPostTheory, type=``l'', lwd =
3)

\subsection{----comparetheory1------------------------------------------------------}\label{comparetheory1}

mean(thetaPostEmp) dtheta = thetas{[}2{]}-thetas{[}1{]} sum(thetas *
densPostTheory * dtheta)

\subsection{----comparetheory2, echo =
FALSE----------------------------------------}\label{comparetheory2-echo-false-}

stopifnot(abs(mean(thetaPostEmp) - sum(thetas * densPostTheory *
dtheta)) \textless{} 1e-3)

\subsection{----mcint---------------------------------------------------------------}\label{mcint}

thetaPostMC = rbeta(n = 1e6, 90, 610) mean(thetaPostMC)

\subsection{\texorpdfstring{----chap2-r-qqplotbeta-1, fig.keep = `high',
fig.cap = ``QQ-plot of our Monte Carlo sample \texttt{thetaPostMC} from
the theoretical distribution and our simulation sample
\texttt{thetaPostEmp}. We could also similarly compare either of these
two distributions to the theoretical distribution function
\texttt{pbeta(.,\ 90,\ 610)}. If the curve lies on the line \(y=x\) this
indicates a good agreement. There are some random differences at the
tails.'', fig.width = 3.5, fig.height =
3.5----}{----chap2-r-qqplotbeta-1, fig.keep = high, fig.cap = QQ-plot of our Monte Carlo sample thetaPostMC from the theoretical distribution and our simulation sample thetaPostEmp. We could also similarly compare either of these two distributions to the theoretical distribution function pbeta(., 90, 610). If the curve lies on the line y=x this indicates a good agreement. There are some random differences at the tails., fig.width = 3.5, fig.height = 3.5----}}\label{chap2-r-qqplotbeta-1-fig.keep-high-fig.cap-qq-plot-of-our-monte-carlo-sample-thetapostmc-from-the-theoretical-distribution-and-our-simulation-sample-thetapostemp.-we-could-also-similarly-compare-either-of-these-two-distributions-to-the-theoretical-distribution-function-pbeta.-90-610.-if-the-curve-lies-on-the-line-yx-this-indicates-a-good-agreement.-there-are-some-random-differences-at-the-tails.-fig.width-3.5-fig.height-3.5-}

qqplot(thetaPostMC, thetaPostEmp, type = ``l'', asp = 1) abline(a = 0, b
= 1, col = ``blue'')

\subsection{----postbeta------------------------------------------------------------}\label{postbeta}

densPost2 = dbeta(thetas, 115, 735) mcPost2 = rbeta(1e6, 115, 735)

sum(thetas * densPost2 * dtheta) \# mean, by numeric integration
mean(mcPost2) \# mean, by MC thetas{[}which.max(densPost2){]} \# MAP
estimate

\subsection{\texorpdfstring{---- roulette-chunk-1, eval = TRUE, echo =
FALSE, fig.keep =
`high'-----}{---- roulette-chunk-1, eval = TRUE, echo = FALSE, fig.keep = high-----}}\label{roulette-chunk-1-eval-true-echo-false-fig.keep-high}

knitr::include\_graphics(`images/roulette.png', dpi = 600)

\subsection{----quantilespost-------------------------------------------------------}\label{quantilespost-}

quantile(mcPost2, c(0.025, 0.975))

\subsection{\texorpdfstring{---- DESeq2-Prediction-Interval, eval =
TRUE, echo = FALSE, fig.show = `hold', fig.keep = `high', fig.cap = ``An
example from @LoveDESeq2 shows plots of the likelihoods (solid lines,
scaled to integrate to 1) and the posteriors (dashed lines) for the
green and purple genes and of the prior (solid black line): due to the
higher dispersion of the purple gene, its likelihood is wider and less
peaked (indicating less information), and the prior has more influence
on its posterior than for the green gene. The stronger curvature of the
green posterior at its maximum translates to a smaller reported standard
error for the MAP logarithmic fold change (LFC) estimate (horizontal
error
bar).''----}{---- DESeq2-Prediction-Interval, eval = TRUE, echo = FALSE, fig.show = hold, fig.keep = high, fig.cap = An example from @LoveDESeq2 shows plots of the likelihoods (solid lines, scaled to integrate to 1) and the posteriors (dashed lines) for the green and purple genes and of the prior (solid black line): due to the higher dispersion of the purple gene, its likelihood is wider and less peaked (indicating less information), and the prior has more influence on its posterior than for the green gene. The stronger curvature of the green posterior at its maximum translates to a smaller reported standard error for the MAP logarithmic fold change (LFC) estimate (horizontal error bar).----}}\label{deseq2-prediction-interval-eval-true-echo-false-fig.show-hold-fig.keep-high-fig.cap-an-example-from-lovedeseq2-shows-plots-of-the-likelihoods-solid-lines-scaled-to-integrate-to-1-and-the-posteriors-dashed-lines-for-the-green-and-purple-genes-and-of-the-prior-solid-black-line-due-to-the-higher-dispersion-of-the-purple-gene-its-likelihood-is-wider-and-less-peaked-indicating-less-information-and-the-prior-has-more-influence-on-its-posterior-than-for-the-green-gene.-the-stronger-curvature-of-the-green-posterior-at-its-maximum-translates-to-a-smaller-reported-standard-error-for-the-map-logarithmic-fold-change-lfc-estimate-horizontal-error-bar.-}

knitr::include\_graphics(c(`images/DESeq2-Prediction-Interval.png'))

\subsection{----callBios------------------------------------------------------------}\label{callbios}

library(``Biostrings'')

\subsection{\texorpdfstring{----BiostringExplore, results = ``hide'',
eval =
FALSE--------------------}{----BiostringExplore, results = hide, eval = FALSE--------------------}}\label{biostringexplore-results-hide-eval-false}

\subsection{GENETIC\_CODE}\label{genetic_code}

\subsection{IUPAC\_CODE\_MAP}\label{iupac_code_map}

\subsection{\texorpdfstring{vignette(package =
``Biostrings'')}{vignette(package = Biostrings)}}\label{vignettepackage-biostrings}

\subsection{\texorpdfstring{vignette(``BiostringsQuickOverview'',
package =
``Biostrings'')}{vignette(BiostringsQuickOverview, package = Biostrings)}}\label{vignettebiostringsquickoverview-package-biostrings}

\subsection{\texorpdfstring{----BiostringCheck, echo=FALSE, results =
``hide''------------------------}{----BiostringCheck, echo=FALSE, results = hide------------------------}}\label{biostringcheck-echofalse-results-hide}

GENETIC\_CODE IUPAC\_CODE\_MAP

\subsection{----BSgenomes-----------------------------------------------------------}\label{bsgenomes}

library(``BSgenome'') ag = available.genomes() length(ag) ag{[}1:2{]}

\subsection{\texorpdfstring{----BSGenomeEcoli,
results=``hide''---------------------------------------}{----BSGenomeEcoli, results=hide---------------------------------------}}\label{bsgenomeecoli-resultshide}

library(``BSgenome.Ecoli.NCBI.20080805'') Ecoli shineDalgarno =
``AGGAGGT'' ecoli = Ecoli\$NC\_010473

\subsection{----window--------------------------------------------------------------}\label{window}

window = 50000 starts = seq(1, length(ecoli) - window, by = window) ends
= starts + window - 1 numMatches = vapply(seq\_along(starts),
function(i) \{ countPattern(shineDalgarno,
ecoli{[}starts{[}i{]}:ends{[}i{]}{]}, max.mismatch = 0) \}, numeric(1))
table(numMatches)

\subsection{\texorpdfstring{----poissonness, fig.keep = `high', fig.cap
= ``Evaluation of a Poisson model for motif counts along the sequence
\textbackslash{}texttt\{Ecoli\$NC\_010473\}.'', fig.width = 4,
fig.height =
5----}{----poissonness, fig.keep = high, fig.cap = Evaluation of a Poisson model for motif counts along the sequence \textbackslash{}texttt\{Ecoli\$NC\_010473\}., fig.width = 4, fig.height = 5----}}\label{poissonness-fig.keep-high-fig.cap-evaluation-of-a-poisson-model-for-motif-counts-along-the-sequence-textttecolinc_010473.-fig.width-4-fig.height-5-}

library(``vcd'') gf = goodfit(numMatches, ``poisson'') summary(gf)
distplot(numMatches, type = ``poisson'')

\subsection{\texorpdfstring{----pattlocIranges1,
results=``hide''-------------------------------------}{----pattlocIranges1, results=hide-------------------------------------}}\label{pattlociranges1-resultshide-}

sdMatches = matchPattern(shineDalgarno, ecoli, max.mismatch = 0)

\subsection{----pattlocIranges2-----------------------------------------------------}\label{pattlociranges2}

betweenmotifs = gaps(sdMatches)

\subsection{\texorpdfstring{----chap2-r-expplotdata-1, fig.keep =
`high', fig.cap = ``Evaluation of fit to the exponential distribution
(black line) of the gaps between the motifs.'', fig.width = 3.5,
fig.height =
4----}{----chap2-r-expplotdata-1, fig.keep = high, fig.cap = Evaluation of fit to the exponential distribution (black line) of the gaps between the motifs., fig.width = 3.5, fig.height = 4----}}\label{chap2-r-expplotdata-1-fig.keep-high-fig.cap-evaluation-of-fit-to-the-exponential-distribution-black-line-of-the-gaps-between-the-motifs.-fig.width-3.5-fig.height-4-}

library(``Renext'') expplot(width(betweenmotifs), rate =
1/mean(width(betweenmotifs)), labels = ``fit'')

\subsection{----gof, echo = FALSE, eval =
FALSE-------------------------------------}\label{gof-echo-false-eval-false-}

\subsection{gofExp.test(width(betweenmotifs))}\label{gofexp.testwidthbetweenmotifs}

\subsection{----chr8HS--------------------------------------------------------------}\label{chr8hs}

library(``BSgenome.Hsapiens.UCSC.hg19'') chr8 = Hsapiens\$chr8 CpGtab =
read.table(``../data/model-based-cpg-islands-hg19.txt'', header = TRUE)
nrow(CpGtab) head(CpGtab) irCpG = with(dplyr::filter(CpGtab, chr ==
``chr8''), IRanges(start = start, end = end))

\subsection{----grCpG---------------------------------------------------------------}\label{grcpg}

grCpG = GRanges(ranges = irCpG, seqnames = ``chr8'', strand = ``+'')
genome(grCpG) = ``hg19''

\subsection{\texorpdfstring{----freqandbayes-ideo, fig.keep = `high',
fig.cap =
``\textbf{\href{https://bioconductor.org/packages/Gviz/}{Gviz}} plot of
CpG locations in a selected region of chromosome 8.'', fig.height =
4----}{----freqandbayes-ideo, fig.keep = high, fig.cap = Gviz plot of CpG locations in a selected region of chromosome 8., fig.height = 4----}}\label{freqandbayes-ideo-fig.keep-high-fig.cap-gviz-plot-of-cpg-locations-in-a-selected-region-of-chromosome-8.-fig.height-4-}

library(``Gviz'') ideo = IdeogramTrack(genome = ``hg19'', chromosome =
``chr8'') plotTracks( list(GenomeAxisTrack(), AnnotationTrack(grCpG,
name = ``CpG''), ideo), from = 2200000, to = 5800000, shape = ``box'',
fill = ``\#006400'', stacking = ``dense'')

\subsection{----CGIview-------------------------------------------------------------}\label{cgiview-}

CGIview =
Views(unmasked(Hsapiens\(chr8), irCpG) NonCGIview = Views(unmasked(Hsapiens\)chr8),
gaps(irCpG))

\subsection{----CGIview2------------------------------------------------------------}\label{cgiview2}

seqCGI = as(CGIview, ``DNAStringSet'') seqNonCGI = as(NonCGIview,
``DNAStringSet'') dinucCpG = sapply(seqCGI, dinucleotideFrequency)
dinucNonCpG = sapply(seqNonCGI, dinucleotideFrequency) dinucNonCpG{[},
1{]} NonICounts = rowSums(dinucNonCpG) IslCounts = rowSums(dinucCpG)

\subsection{----transitions---------------------------------------------------------}\label{transitions}

TI = matrix( IslCounts, ncol = 4, byrow = TRUE) TnI = matrix(NonICounts,
ncol = 4, byrow = TRUE) dimnames(TI) = dimnames(TnI) = list(c(``A'',
``C'', ``G'', ``T''), c(``A'', ``C'', ``G'', ``T''))

\subsection{----MI------------------------------------------------------------------}\label{mi}

MI = TI /rowSums(TI) MI MN = TnI / rowSums(TnI) MN

\subsection{----STATI---------------------------------------------------------------}\label{stati}

freqIsl = alphabetFrequency(seqCGI, baseOnly = TRUE, collapse =
TRUE){[}1:4{]} freqIsl / sum(freqIsl) freqNon =
alphabetFrequency(seqNonCGI, baseOnly = TRUE, collapse = TRUE){[}1:4{]}
freqNon / sum(freqNon)

\subsection{\texorpdfstring{---- book-chunk-1, eval = TRUE, echo =
FALSE, fig.keep =
`high'---------}{---- book-chunk-1, eval = TRUE, echo = FALSE, fig.keep = high---------}}\label{book-chunk-1-eval-true-echo-false-fig.keep-high}

knitr::include\_graphics(`images/book\_icon.png', dpi = 400)

\subsection{----alphabeta-----------------------------------------------------------}\label{alphabeta}

alpha = log((freqIsl/sum(freqIsl)) / (freqNon/sum(freqNon))) beta =
log(MI / MN)

\subsection{----scorepatt-----------------------------------------------------------}\label{scorepatt}

x = ``ACGTTATACTACG'' scorefun = function(x) \{ s = unlist(strsplit(x,
``'')) score = alpha{[}s{[}1{]}{]} if (length(s) \textgreater{}= 2) for
(j in 2:length(s)) score = score + beta{[}s{[}j-1{]}, s{[}j{]}{]} score
\} scorefun(x)

\subsection{----scorefun1-----------------------------------------------------------}\label{scorefun1}

generateRandomScores = function(s, len = 100, B = 1000) \{ alphFreq =
alphabetFrequency(s) isGoodSeq = rowSums(alphFreq{[},
5:ncol(alphFreq){]}) == 0 s = s{[}isGoodSeq{]} slen = sapply(s, length)
prob = pmax(slen - len, 0) prob = prob / sum(prob) idx =
sample(length(s), B, replace = TRUE, prob = prob) ssmp = s{[}idx{]}
start = sapply(ssmp, function(x) sample(length(x) - len, 1)) scores =
sapply(seq\_len(B), function(i)
scorefun(as.character(ssmp{[}{[}i{]}{]}{[}start{[}i{]}+(1:len){]})) )
scores / len \} scoresCGI = generateRandomScores(seqCGI) scoresNonCGI =
generateRandomScores(seqNonCGI)

\subsection{\texorpdfstring{----chap2-r-ScoreMixture-1, fig.keep =
`high', fig.cap = ``Island and non-island scores as generated by the
function \texttt{generateRandomScores}. This is the first instance of a
\textbf{mixture} we encounter. We will revisit them in ChapterÂ
\textbackslash{}@ref(Chap:Mixtures).'', fig.height =
5----}{----chap2-r-ScoreMixture-1, fig.keep = high, fig.cap = Island and non-island scores as generated by the function generateRandomScores. This is the first instance of a mixture we encounter. We will revisit them in ChapterÂ \textbackslash{}@ref(Chap:Mixtures)., fig.height = 5----}}\label{chap2-r-scoremixture-1-fig.keep-high-fig.cap-island-and-non-island-scores-as-generated-by-the-function-generaterandomscores.-this-is-the-first-instance-of-a-mixture-we-encounter.-we-will-revisit-them-in-chapteruxe2-refchapmixtures.-fig.height-5-}

br = seq(-0.6, 0.7, length.out = 50) h1 = hist(scoresCGI, breaks = br,
plot = FALSE) h2 = hist(scoresNonCGI, breaks = br, plot = FALSE)
plot(h1, col = rgb(0, 0, 1, 1/4), xlim = c(-0.5, 0.5), ylim=c(0,120))
plot(h2, col = rgb(1, 0, 0, 1/4), add = TRUE)

\subsection{----savescoresforChap4, echo = FALSE,
eval=FALSE------------------------}\label{savescoresforchap4-echo-false-evalfalse}

\subsection{\#\#\#This is for provenance reasons, keep track of how the
data}\label{this-is-for-provenance-reasons-keep-track-of-how-the-data}

\subsection{\#\#\#were generated for the EM exercise in Chapter
4.}\label{were-generated-for-the-em-exercise-in-chapter-4.}

\subsection{Mdata=c(scoresCGI,scoresNonCGI)}\label{mdatacscorescgiscoresnoncgi}

\subsection{MM1=sample(Mdata{[}1:1000{]},800)}\label{mm1samplemdata11000800}

\subsection{MM2=sample(Mdata{[}1001:2000{]},1000)}\label{mm2samplemdata100120001000}

\subsection{Myst=c(MM1,MM2);names(Myst)=NULL}\label{mystcmm1mm2namesmystnull}

\subsection{\texorpdfstring{saveRDS(c(MM1,MM2),``../data/Myst.rds'')}{saveRDS(c(MM1,MM2),../data/Myst.rds)}}\label{saverdscmm1mm2..datamyst.rds}

\subsection{\#\#\#True value of m1,m2,s1 and
s2}\label{true-value-of-m1m2s1-and-s2}

\subsection{}\label{section}

\subsection{----checkhists, echo =
FALSE--------------------------------------------}\label{checkhists-echo-false}

stopifnot(max(h1\(counts) < 120, max(h2\)counts) \textless{} 120,
h1\(breaks[1] >= br[1], h1\)breaks{[}length(h1\$breaks){]} \textless{}=
br{[}length(br){]},
h2\(breaks[1] >= br[1], h2\)breaks{[}length(h2\$breaks){]} \textless{}=
br{[}length(br){]})

mtb = read.table(``../data/M\_tuberculosis.txt'', header = TRUE)
head(mtb, n = 4)

\subsection{----ProlMyc-------------------------------------------------------------}\label{prolmyc-}

pro = mtb{[} mtb\$AmAcid == ``Pro'', ``Number''{]} pro/sum(pro)

staph = readDNAStringSet(``../data/staphsequence.ffn.txt'', ``fasta'')

\subsection{----staphex-------------------------------------------------------------}\label{staphex-}

staph{[}1:3, {]} staph

\subsection{----GCfreq--------------------------------------------------------------}\label{gcfreq}

letterFrequency(staph{[}{[}1{]}{]}, letters = ``ACGT'', OR = 0) GCstaph
= data.frame( ID = names(staph), GC =
rowSums(alphabetFrequency(staph){[}, 2:3{]} / width(staph)) * 100 )

\subsection{\texorpdfstring{----chap2-r-SlidingGC-1, fig.keep = `high',
fig.cap = ``GC content along sequence 364 of the \emph{Staphylococcus
Aureus} genome.'', fig.width =
7----}{----chap2-r-SlidingGC-1, fig.keep = high, fig.cap = GC content along sequence 364 of the Staphylococcus Aureus genome., fig.width = 7----}}\label{chap2-r-slidinggc-1-fig.keep-high-fig.cap-gc-content-along-sequence-364-of-the-staphylococcus-aureus-genome.-fig.width-7-}

window = 100 gc = rowSums(
letterFrequencyInSlidingView(staph{[}{[}364{]}{]}, window,
c(``G'',``C'')))/window plot(x = seq(along = gc), y = gc, type = ``l'')

\subsection{\texorpdfstring{----chap2-r-SmoothSlidingGC-1, fig.keep =
`high', fig.cap = ``Smoothed GC content along sequence 364 of the
\emph{Staphylococcus Aureus} genome.'', fig.width =
7----}{----chap2-r-SmoothSlidingGC-1, fig.keep = high, fig.cap = Smoothed GC content along sequence 364 of the Staphylococcus Aureus genome., fig.width = 7----}}\label{chap2-r-smoothslidinggc-1-fig.keep-high-fig.cap-smoothed-gc-content-along-sequence-364-of-the-staphylococcus-aureus-genome.-fig.width-7-}

plot(x = seq(along = gc), y = gc, type = ``l'') lines(lowess(x =
seq(along = gc), y = gc, f = 0.2), col = 2)

\subsection{----histobeta4, fig.width = 3.5, fig.height =
3.5-----------------------}\label{histobeta4-fig.width-3.5-fig.height-3.5}

theta = thetas{[}1:500{]} dfbetas = data.frame(theta, db1=
dbeta(theta,0.5,0.5), db2= dbeta(theta,1,1), db3= dbeta(theta,10,30),
db4 = dbeta(theta, 20, 60), db5 = dbeta(theta, 50, 150))
require(reshape2) datalong = melt(dfbetas, id=``theta'')
ggplot(datalong) + geom\_line(aes(x = theta,y=value,colour=variable)) +
theme(legend.title=element\_blank()) + geom\_vline(aes(xintercept=0.25),
colour=``\#990000'', linetype=``dashed'')+ scale\_colour\_discrete(name
=``Prior'', labels=c(``B(0.5,0.5)'',``U(0,1)=B(1,1)'',``B(10,30)'',
``B(20,60)'',``B(50,150)''))


\end{document}
